File: prep/Rcode/hike.R
# # connect to stdin to get data
# f <- file("stdin")
# on.exit(close(f))
# # read input
# input <- strsplit(readLines(f)," ")
# # create variables
# nSteps <- as.numeric(strsplit(input[[1]]," "))
# stepString <- as.numeric(strsplit(input[[2]]," "))

nSteps = 8
stepString = "UDDDUDUU"
path <- 0
valleyCount <- 0

for(i in 1:nSteps){
  # create path based on steps
  step <- substr(stepString,i,i)
  if(step == "U"){
    path <- c(path, path[i] + 1)
  }
  if(step == "D"){
    path <- c(path, path[i] - 1)
  }
  
  # check if new valley exists 
  if(sign(path[i]) >= 0 && sign(path[(i + 1)]) < 0 ){
    valleyCount = valleyCount + 1
  }
}

write(valleyCount,stdout())

File: prep/py_code/hike.py
import array as arr

n_steps = 8
step_string = "UDDDUDUU"


def hike(n_steps, step_string):
    path = arr.array('i', [0])
    valley_count = 0
    # create path based on steps
    for i in range(n_steps):
        step = step_string[i]
        if step == "U":
            path.append(path[i] + 1)

        if step == "D":
            path.append(path[i] - 1)

        # check if new valley exists
        if path[i] >= 0:
            if path[(i + 1)] < 0:
                valley_count = valley_count + 1

    return valley_count



print(hike(n_steps, step_string))
